lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+                             numCols: Int): Unit = {
+    this.numCols = numCols
+    labelsArray = new FloatSwigArray(this.rowCount)
+    weightChunkedArrayOpt.foreach { _ =>
+      weightArrayOpt = Some(new FloatSwigArray(this.rowCount)) }
+    initScoreChunkedArrayOpt.foreach{ _ =>
+      initScoreArrayOpt = Some(new DoubleSwigArray(this.initScoreCount)) }
+    featuresArray = new DoubleSwigArray(numCols * this.rowCount)
+    groupColumnValuesArray = new Array[Row](this.rowCount.toInt)
+  }
+
+  private def parallelInitializeRows(weightChunkedArrayOpt: Option[FloatChunkedArray],
+                                     initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+                                     numCols: Int): Unit = {
+    // Initialize arrays if they are not defined - first thread to get here does the initialization for all of them
+    if (labelsArray == null) {
+      this.synchronized {
+        if (labelsArray == null) {
+          initializeRows(weightChunkedArrayOpt, initScoreChunkedArrayOpt, numCols)
+        }
+      }
+    }
+  }
+
+  private def parallelizedCopy(labelsChunkedArray: FloatChunkedArray,
+                               weightChunkedArrayOpt: Option[FloatChunkedArray],
+                               initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+                               featuresChunkedArray: DoubleChunkedArray,
+                               groupColumnValues: ListBuffer[Row],
+                               numCols: Int): Unit = {
+    // Parallelized copy to common arrays
+    var threadRowStartIndex = 0L
+    var threadInitScoreStartIndex = 0L
+    var threadIndexesStartIndex = 0L
+    var threadIndPtrStartIndex = 0L
+    this.synchronized {
+      val labelsSize = labelsChunkedArray.getAddCount()
+      threadRowStartIndex = this.threadRowStartIndex
+      this.threadRowStartIndex = this.threadRowStartIndex + labelsSize.toInt
+
+      threadInitScoreStartIndex = this.threadInitScoreStartIndex
+      val initScoreSize = initScoreChunkedArrayOpt.map(_.getAddCount())
+      initScoreSize.foreach(size => this.threadInitScoreStartIndex = this.threadInitScoreStartIndex + size)
+    }
+    ChunkedArrayUtils.copyChunkedArray(labelsChunkedArray, this.labelsArray, threadRowStartIndex, chunkSize)
+    weightChunkedArrayOpt.foreach {
+      weightChunkedArray =>
+        ChunkedArrayUtils.copyChunkedArray(weightChunkedArray, this.weightArrayOpt.get, threadRowStartIndex,
+          chunkSize)
+    }
+    initScoreChunkedArrayOpt.foreach {
+      initScoreChunkedArray =>
+        ChunkedArrayUtils.copyChunkedArray(initScoreChunkedArray, this.initScoreArrayOpt.get,
+          threadInitScoreStartIndex, chunkSize)
+    }
+    ChunkedArrayUtils.copyChunkedArray(featuresChunkedArray, this.featuresArray, threadRowStartIndex * numCols,
+      chunkSize)
+    for (index <- threadRowStartIndex until threadRowStartIndex + groupColumnValues.length) {
```

Comment:
Can this `for` loop be replaced by a `copyToArray` call with `threadRowStartIndex` as the `start` parameter?

Reply:
done!

Reply:
I think replacing this with copyToArray will solve the O(N^2) issue @TomFinley mentioned above.